check_vars:  Test if host_port is in use for check_docker

kernelzilla opened the pull request:
##### SUMMARY

Fail check_vars when the defined host_port is in use. Ref issue related #3956.

##### ISSUE TYPE
 - Bugfix Pull Request


##### COMPONENT NAME
 - Installer

##### AWX VERSION
<!--- Paste verbatim output from `make VERSION` between quotes below -->
```
awx: 4.0.0
```

Commit:
check_vars:  Test if host_port is in use for check_docker

Fail check_vars when the defined host_port is in use. Ref issue #3956.

ryanpetrello commented on installer/roles/check_vars/tasks/check_docker.yml:
```diff
@@ -12,3 +12,10 @@
+    port: "{{ host_port }}"
+    timeout: 2
+    state: stopped
+    msg: "host_port is currently in use"
```
Could we make this error message clearer, something more like a generic socket bind error:

```
~ nc -kl 8043
Ncat: bind to :::8043: Address already in use. QUITTING.
~ python3 -m http.server 8043
...
OSError: [Errno 98] Address already in use
```

(might be nice to actually show the port number)

shanemcd reviewed (commented):
Will this fail even if AWX itself is what is already running on the host? Does anything stop the services beforehand? Would be good to test this out with an existing installation.

Commit:
check_vars:  confirm host_port listening process in check_docker

Update the check for processes listening on host_port to include
the name and exclude docker-proxy when the listening container
is awx_web to allow upgrade path.

Tested in the following scenarios:

- Clean install with no services listening on host_port

- Clean install with a service listening on host_port

- Existing install with awx_web container listening on host_port

See related #3956.

kernelzilla commented:
Thank you for the feedback.  I've tested and pushed a new commit to address the existing awx_web process scenario as well as provide the name of the listening process name if host_port is in use.

shanemcd commented:
Hello and I'm sorry that we dropped the ball on reviewing this PR for so long. I just tried this out locally and I am seeing this:

```
TASK [check_vars : gather process name listening on host_port] ***************************************************************************************
changed: [localhost] => changed=true
  cmd: ps hp `fuser -n tcp "80" 2> /dev/null | cut -d ' ' -f 2` -o comm
  delta: '0:00:00.015846'
  end: '2019-11-12 11:52:30.189882'
  failed_when_result: false
  msg: non-zero return code
  rc: 1
  start: '2019-11-12 11:52:30.174036'
  stderr: |-
    error: process ID list syntax error

    Usage:
     ps [options]

     Try 'ps --help <simple|list|output|threads|misc|all>'
      or 'ps --help <s|l|o|t|m|a>'
     for additional help text.

    For more details see ps(1).
  stderr_lines: <omitted>
  stdout: ''
  stdout_lines: <omitted>
```

Which does not cause the playbook to fail since we ignore errors, but it does not seem like the intended behavior. I will leave this open for a bit to give you the chance to circle back if you want to, or we can just close it since no one else has reported this.

ryanpetrello commented:
Closing due to age.
